infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesDeployments.java

```diff
@@ -1168,4 +1171,24 @@ public void wait(long timeout, TimeUnit timeUnit)
       }
     }
   }
+
+  @VisibleForTesting
+  void addPullSecretsOfSA(Deployment deployment) throws InfrastructureException {
+    final PodSpec podSpec = deployment.getSpec().getTemplate().getSpec();
+    final String podServiceAccountName = podSpec.getServiceAccountName();
+    if (podServiceAccountName == null) {
+      return;
+    }
+    ServiceAccount deploymentServiceAccount =
+        clientFactory
+            .create(workspaceId)
+            .serviceAccounts()
+            .inNamespace(namespace)
+            .withName(podServiceAccountName)
+            .get();
+    if (deploymentServiceAccount == null) {
+      return;
+    }
+    podSpec.getImagePullSecrets().addAll(deploymentServiceAccount.getImagePullSecrets());
```

Comment:
can `deploymentServiceAccount.getImagePullSecrets()` return null ?

Reply:
from: 
https://github.com/fabric8io/kubernetes-client/blob/d1740ffeaa624bd63b6d929b25d764a477c1c17e/kubernetes-model-generator/kubernetes-model-core/src/generated/java/io/fabric8/kubernetes/api/model/PodSpec.java#L97
I assumed not

Reply:
does it actually work? If this https://github.com/eclipse-che/che-server/blob/6670c6986b1cb3da5e733e1a0b48db78b393ee3d/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java#L168 sets the list as `ImmutableList`, I would imagine `addAll()` won't work. IMHO since there is a setter `setImagePullSecrets`, I don't think we can rely on anything here. You will need to create new list here, put together `podSpec.getImagePullSecrets()` and `deploymentServiceAccount.getImagePullSecrets()` with nullchecks and ideally get rid of duplicates. Only then set it to `podSpec.setImagePullSecrets(...)`

Reply:
![image](https://user-images.githubusercontent.com/73115616/127158016-fa0bc95e-34fe-4e04-ad03-e0618b8b70d2.png)
I tried debugging and in looks like normal ArrayList. Maybe we are overriding the ImutableList somewhere or I am missing something?
![image](https://user-images.githubusercontent.com/73115616/127158522-50ad2172-9ce7-4509-9070-d81b83d012cf.png)
and this after the method

Reply:
ok, but still the PodSpec object definition allows easily to set it to something different than `ArrayList`. This might bite us in the future. Please implement it more defensively.

Reply:
yea, I agree, will do